internal/controller/nimpipeline_controller.go

```diff
@@ -33,30 +39,257 @@ type NIMPipelineReconciler struct {
+	err := r.Get(ctx, types.NamespacedName{Name: nimService.Name, Namespace: nimService.Namespace}, existingService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Creating new NIMService", "name", nimService.Name)
+			return r.Create(ctx, nimService)
+		}
+		logger.Error(err, "Failed to get NIMService", "name", nimService.Name)
+		return err
+	}
+
+	// Update the existing NIMService's spec if needed
+	logger.Info("Updating existing NIMService", "name", nimService.Name)
+	existingService.Spec = nimService.Spec
+	if err := r.Update(ctx, existingService); err != nil {
+		logger.Error(err, "Failed to update NIMService", "name", nimService.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (r *NIMPipelineReconciler) cleanupDisabledNIMs(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline, enabledServices map[string]bool) error {
+	logger := log.FromContext(ctx)
+
+	serviceList := &appsv1alpha1.NIMServiceList{}
+	if err := r.List(ctx, serviceList, client.InNamespace(nimPipeline.Namespace)); err != nil {
+		logger.Error(err, "Failed to list NIMServices")
+		return err
+	}
+
+	for _, svc := range serviceList.Items {
+		// Cleanup any stale NIM services if they are part of the pipeline but are disabled
+		if enabled, exists := enabledServices[svc.Name]; exists && !enabled {
+			if err := r.deleteService(ctx, &svc); err != nil {
+				logger.Error(err, "Unable to delete disabled NIM service", "Name", svc.Name)
+				return err
+			}
+			continue
+		}
+	}
+	return nil
+}
+
+func (r *NIMPipelineReconciler) updateStatus(ctx context.Context, nimPipeline *appsv1alpha1.NIMPipeline, enabledServices map[string]bool) error {
+	logger := log.FromContext(ctx)
+
+	serviceList := &appsv1alpha1.NIMServiceList{}
+	if err := r.List(ctx, serviceList, client.InNamespace(nimPipeline.Namespace)); err != nil {
+		logger.Error(err, "Failed to list NIMServices")
+		return err
+	}
+
+	// Default overall state to "NotReady"
+	overallState := appsv1alpha1.NIMServiceStatusNotReady
+	serviceStates := make(map[string]string)
+	allServicesReady := true
+
+	for _, svc := range serviceList.Items {
+		if enabled, exists := enabledServices[svc.Name]; !exists || !enabled {
+			continue
+		}
+
+		// Update service states
+		serviceStates[svc.Name] = svc.Status.State
+
+		switch svc.Status.State {
+		case appsv1alpha1.NIMServiceStatusFailed:
+			// If any service has failed, the overall state should be "Failed"
+			overallState = appsv1alpha1.NIMServiceStatusFailed
+			allServicesReady = false
+		case appsv1alpha1.NIMServiceStatusNotReady:
+			fallthrough
+		case appsv1alpha1.NIMServiceStatusPending:
+			// If any service is not ready, mark overall readiness as false
+			allServicesReady = false
+		}
+	}
+
+	// If all services are ready and no failures were detected, set overall state to "Ready"
+	if allServicesReady && overallState != appsv1alpha1.NIMServiceStatusFailed {
+		overallState = appsv1alpha1.NIMServiceStatusReady
+	}
```

Comment:
Can we copy over the condition of the NIMService to the NIMPipeline, that will give user better picture of whats going on?
https://github.com/NVIDIA/k8s-nim-operator/blob/5a9702b6918fd5dc72b4b0845a3059591c056251/internal/conditions/conditions.go#L111

Reply:
We cannot add same conditiontype from all services, but need to collect a cumulative reason with single type. Will handle condition updates as a separate PR.